test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/PreserveTypesAndAttributesTests.cs

```diff
@@ -0,0 +1,395 @@
+                Assert.Equal(result1[i].Name, result2[i].Name);
+            }
+            Assert.Equal(result3.Count, result2.Count);
+            for (var i = 0; i < result1.Count; i++)
+            {
+                Assert.Equal(result3[i].Id, result2[i].Id);
+                Assert.Equal(result3[i].CorporationName, result2[i].CorporationName);
+                Assert.Equal(result3[i].City, result2[i].City);
+                Assert.Equal(result3[i].Name, result2[i].Name);
+            }
+        }
+    }
+
+
+    // One level of inheritance classes
+
+    public interface ICustomer
+    {
+        string Id { get; set; }
+        string City { get; set; }
+        string CorporationName { get; set; }
+        string Name { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class OneLevelCustomer1 : ICustomer
+    {
+        [OriginalName("CustomerID")]
+        public string Id { get; set; }
+
+        public string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public string Name { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class BaseCustomer
+    {
+        [OriginalName("CustomerID")]
+        public string Id { get; set; }
+    }
+
+    public class OneLevelCustomer2 : BaseCustomer
+    {
+        public string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public string Name { get; set; }
+    }
+
+
+    // More than one level of inheritance classes
+
+    public abstract class ACustomer : ICustomer
+    {
+        public virtual string Id { get; set; }
+        public virtual string City { get; set; }
+        public virtual string CorporationName { get; set; }
+        public virtual string Name { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class TwoLevelCustomer1 : ACustomer
+    {
+        [OriginalName("CustomerID")]
+        public override string Id { get; set; }
+
+        public override string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public override string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public override string Name { get; set; }
+    }
+
+    public interface ICustomer2 : ICustomer
+    {
+        List<Order> OrdersList { get; set; }
+    }
+
+    [Key("CustomerID")]
+    public class TwoLevelCustomer2 : ICustomer2
+    {
+        [OriginalName("CustomerID")]
+        public string Id { get; set; }
+
+        public string City { get; set; }
+
+        [OriginalName("CompanyName")]
+        public string CorporationName { get; set; }
+
+        [OriginalName("ContactName")]
+        public string Name { get; set; }
+
+        [OriginalName("Orders")]
+        public List<Order> OrdersList { get; set; }
+    }
+
+    [Key("OrderID")]
+    public class Order
```

Comment:
Since these model class names are common, it may be good to consider moving them to inner classes or a namespace where they are less likely to cause conflict.